src/main/java/com/commercetools/sync/producttypes/helpers/BatchProcessor.java

```diff
@@ -0,0 +1,154 @@
+                        handleError(invalidProductTypeDraftException);
+                    }
+                } else {
+                    final String errorMessage = format(PRODUCT_TYPE_DRAFT_KEY_NOT_SET, productTypeDraft.getName());
+                    handleError(new InvalidProductTypeDraftException(errorMessage));
+                }
+            } else {
+
+                handleError(new InvalidProductTypeDraftException(PRODUCT_TYPE_DRAFT_IS_NULL));
+            }
+        }
+    }
+
+    @Nonnull
+    private static Set<String> getReferencedProductTypeKeys(@Nonnull final ProductTypeDraft productTypeDraft)
+        throws InvalidProductTypeDraftException {
+
+        final Set<String> referencedProductTypeKeys = new HashSet<>();
+        final List<AttributeDefinitionDraft> attributeDefinitionDrafts = productTypeDraft.getAttributes();
+        if (attributeDefinitionDrafts == null) {
+            return emptySet();
+        }
+
+
+        final List<String> invalidAttributeDefinitionNames = new ArrayList<>();
+
+        for (AttributeDefinitionDraft attributeDefinitionDraft : attributeDefinitionDrafts) {
+            if (attributeDefinitionDraft != null) {
+                final AttributeType attributeType = attributeDefinitionDraft.getAttributeType();
+                try {
+                    getProductTypeKey(attributeType).ifPresent(referencedProductTypeKeys::add);
+                } catch (InvalidReferenceException invalidReferenceException) {
+                    invalidAttributeDefinitionNames.add(attributeDefinitionDraft.getName());
+
+                }
+            }
+        }
+
+        if (!invalidAttributeDefinitionNames.isEmpty()) {
+            final String errorMessage = format(PRODUCT_TYPE_HAS_INVALID_REFERENCES, productTypeDraft.getKey(),
+                invalidAttributeDefinitionNames);
+            throw new InvalidProductTypeDraftException(errorMessage,
+                new InvalidReferenceException(BLANK_ID_VALUE_ON_RESOURCE_IDENTIFIER));
+        }
+
+        return referencedProductTypeKeys;
+    }
+
+    @Nonnull
+    private static Optional<String> getProductTypeKey(@Nonnull final AttributeType attributeType)
+        throws InvalidReferenceException {
+
+        if (attributeType instanceof NestedAttributeType) {
+            final NestedAttributeType nestedElementType = (NestedAttributeType) attributeType;
+            return Optional.of(getProductTypeKey(nestedElementType));
+        } else if (attributeType instanceof SetAttributeType) {
+            final SetAttributeType setAttributeType = (SetAttributeType) attributeType;
+            return getProductTypeKey(setAttributeType.getElementType());
+        }
+        return Optional.empty();
+    }
+
+    @Nonnull
+    private static String getProductTypeKey(@Nonnull final NestedAttributeType nestedAttributeType)
+        throws InvalidReferenceException {
+
+        final String key = nestedAttributeType.getTypeReference().getId();
```

Comment:
yes and also it is a required value for a nested type on the API so it should not be passed `null`. The sync library cannot null check everything which is already stated by the API and SDK contract that it should not be `null`. Otherwise the codebase will be overcomplicated with exessive null checks.

Reply:
I asked this because in this case, it's a potential NPE and it looks like the same problem if there is no ID set, isn't it? because if there is no type reference set you can return the same checked exception instead of blocking all batch processor?

Reply:
totally valid.